### ISSUE TITLE
Exception in GemmParams when using GravesLSTM and TruncatedBPTT

### ISSUE THREAD

**Issue by ericson3000** [deeplearning4j/deeplearning4j#1204]
The Matrix "c" is not in c-order it is in "f"-order. Stack trace is as follows:

`Exception in thread "main" java.lang.IllegalArgumentException: c must be f order, offset 0 and have length == c.data.length
	at org.nd4j.linalg.api.blas.params.GemmParams.<init>(GemmParams.java:59)
	at org.nd4j.linalg.api.blas.impl.BaseLevel3.gemm(BaseLevel3.java:78)
	at org.nd4j.linalg.factory.Nd4j.gemm(Nd4j.java:713)
	at org.deeplearning4j.nn.layers.recurrent.GravesLSTM.activateHelper(GravesLSTM.java:410)
	at org.deeplearning4j.nn.layers.recurrent.GravesLSTM.rnnActivateUsingStoredState(GravesLSTM.java:533)
	at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.rnnActivateUsingStoredState(MultiLayerNetwork.java:2242)
	at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.computeGradientAndScore(MultiLayerNetwork.java:1748)
	at org.deeplearning4j.optimize.solvers.BaseOptimizer.gradientAndScore(BaseOptimizer.java:132)
	at org.deeplearning4j.optimize.solvers.BaseOptimizer.optimize(BaseOptimizer.java:151)
	at org.deeplearning4j.optimize.Solver.optimize(Solver.java:52)
	at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.doTruncatedBPTT(MultiLayerNetwork.java:1299)
	at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.fit(MultiLayerNetwork.java:1157)
	at eric.masterproject.dl4j.lstmView.LSTMAnalysis.trainAndTestModel(LSTMAnalysis.java:93)
	at eric.masterproject.dl4j.lstmView.LSTMAnalysis.main(LSTMAnalysis.java:42)`

The following Configuration was used:

`		MultiLayerConfiguration conf = new NeuralNetConfiguration.Builder()
				.learningRate(0.1)
				.iterations(20)
				.seed(123)
				.list(2)
				.layer(0, new GravesLSTM.Builder().nIn(FEATURES).nOut(LABELS)
						.updater(Updater.ADAGRAD)
						.activation("tanh").weightInit(WeightInit.UNIFORM)
						.dist(new UniformDistribution(-0.08, 0.08))
						.build())
				.layer(1, new RnnOutputLayer.Builder(LossFunction.XENT).activation("tanh")        
						.updater(Updater.ADAGRAD)
						.nIn(LABELS).nOut(LABELS).weightInit(WeightInit.UNIFORM)
						.dist(new UniformDistribution(-0.08, 0.08))
						.build())
				.backpropType(BackpropType.TruncatedBPTT)
				.tBPTTBackwardLength(10)
				.tBPTTForwardLength(10)
				.pretrain(false)
				.backprop(true)
				.build();`

**Comment by AlexDBlack** [comment]
Can you post the version you are using? Both DL4J and ND4J.

**Comment by ericson3000** [comment]
<nd4j.version>0.4-rc3.8</nd4j.version>
        <dl4j.version>	0.4-rc3.8</dl4j.version>

**Comment by AlexDBlack** [comment]
I haven't seen this in practice before (and I've used LSTMs a lot). Can you give me something I can run to reproduce this? The configuration above isn't enough.

**Comment by ericson3000** [comment]
package eric.masterproject.dl4j.lstmView;

import java.io.File;
import java.io.IOException;

import org.canova.api.records.reader.SequenceRecordReader;
import org.canova.api.records.reader.impl.CSVSequenceRecordReader;
import org.canova.api.split.FileSplit;
import org.deeplearning4j.datasets.canova.SequenceRecordReaderDataSetIterator;
import org.deeplearning4j.datasets.canova.SequenceRecordReaderDataSetIterator.AlignmentMode;
import org.deeplearning4j.datasets.iterator.DataSetIterator;
import org.deeplearning4j.nn.conf.BackpropType;
import org.deeplearning4j.nn.conf.MultiLayerConfiguration;
import org.deeplearning4j.nn.conf.NeuralNetConfiguration;
import org.deeplearning4j.nn.conf.Updater;
import org.deeplearning4j.nn.conf.distribution.UniformDistribution;
import org.deeplearning4j.nn.conf.layers.GravesLSTM;
import org.deeplearning4j.nn.conf.layers.RnnOutputLayer;
import org.deeplearning4j.nn.multilayer.MultiLayerNetwork;
import org.deeplearning4j.nn.weights.WeightInit;
import org.deeplearning4j.optimize.listeners.ScoreIterationListener;
import org.nd4j.linalg.lossfunctions.LossFunctions.LossFunction;

/**
 * LSTM Test
 */
public class LSTMAnalysis {

	private final static int FEATURES = 5;
	private final static int LABELS = 1;
	private final static double TRAIN_TEST_RATIO = 0.7;

	/**
	 * 
	 * @param args
	 * @throws IOException
	 * @throws InterruptedException
	 */
	public static void main(String[] args) throws IOException, InterruptedException {

		final MultiLayerNetwork model = trainAndTestModel();
		
//		final ComplexChart2 charting = new ComplexChart2("LSTM/RNN-Model", model, samples);
//		charting.pack();
//		RefineryUtilities.centerFrameOnScreen(charting);
//		charting.setVisible(true);
	}

	/**
	 * 
	 * @param featuresAndLabels
	 * @param timePoint
	 */
	private static MultiLayerNetwork trainAndTestModel() {

		// 1. setup the model configuration
		MultiLayerConfiguration conf = new NeuralNetConfiguration.Builder()
//				.optimizationAlgo(OptimizationAlgorithm.STOCHASTIC_GRADIENT_DESCENT)
				.learningRate(0.1)
				.iterations(20)
//				.rmsDecay(0.95)
				.seed(123)
//				.regularization(true).l2(0.01)
				.list(2)
				.layer(0, new GravesLSTM.Builder().nIn(FEATURES).nOut(LABELS)
						.updater(Updater.ADAGRAD)
						.activation("tanh").weightInit(WeightInit.UNIFORM)
						.dist(new UniformDistribution(-0.08, 0.08))
						.build())
				.layer(1, new RnnOutputLayer.Builder(LossFunction.XENT).activation("tanh")        //MCXENT + softmax for classification
						.updater(Updater.ADAGRAD)
						.nIn(LABELS).nOut(LABELS).weightInit(WeightInit.UNIFORM)
						.dist(new UniformDistribution(-0.08, 0.08))
						.build())
//				.backpropType(BackpropType.TruncatedBPTT)
//				.tBPTTBackwardLength(FEATURES)
//				.tBPTTForwardLength(FEATURES)
				.pretrain(false)
				.backprop(true)
				.build();
		
		// 3. instantiate the configurred network
		final MultiLayerNetwork model = new MultiLayerNetwork(conf);
		model.init();

		// 4. set up score listeners (optional)
//		model.setListeners(new ScoreIterationListener(10), 
//		new HistogramIterationListener(10));
		model.setListeners(new ScoreIterationListener(2));
		 
		try {
			model.fit(getIterator());
		} catch (IOException | InterruptedException e) {
			// TODO Auto-generated catch block
			System.out.println("ERROR while creating the Data iterator!");
			e.printStackTrace();
		}
		return model;
	}
	
	/**
	 * 
	 * @return
	 * @throws IOException
	 * @throws InterruptedException
	 */
	private static DataSetIterator getIterator() throws IOException, InterruptedException{
		
		final File dataRootDir = new File("/Users/eric/Desktop/TestData/");
		final String[] fileFormats = {"csv"};
		final String dataSeperator = ",";
		
		//use 1 to skip the header
		final int firstDataRow = 1;
		
		//use 1 so each file is an DataSet, if there are 3 files in our root-dir and numMiniBatch is 3 
		//we will get one DataSet that contains all timeseries
		final int numMiniBatches = 1;
		
		final SequenceRecordReader featureReader = new CSVSequenceRecordReader(firstDataRow, dataSeperator);
        featureReader.initialize(new FileSplit(dataRootDir, fileFormats));
        
        //start the label-reader one row after feature-reader so the label represents the next timestep value
        final SequenceRecordReader lableReader = new CSVSequenceRecordReader(firstDataRow+1, dataSeperator);
        lableReader.initialize(new FileSplit(dataRootDir, fileFormats));
        
        //Because we do regression=true the number of Labels is not used and can be set to anything (here =0). 
        //AlignmentMode.ALIGN_START is needed because our label-series is one timestep shorter than our feature-series.
        //It will move the first label up to the first feature and fill missing trailing values with 0's
        final SequenceRecordReaderDataSetIterator iter =
                new SequenceRecordReaderDataSetIterator(featureReader, lableReader, numMiniBatches, 0, true, AlignmentMode.ALIGN_START);

		return iter;
	}
	
}

**Comment by agibsonccc** [comment]
Is this still the case?

**Comment by AlexDBlack** [comment]
Just ran a modified version of this locally on latest version. Not seeing any issues.
